cmake/kconfig.cmake

```diff
@@ -5,6 +5,22 @@
 file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/kconfig/include/generated)
 file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/kconfig/include/config)
 
+# Support multiple SOC_ROOT
+set(WRITE WRITE)
+foreach(root ${SOC_ROOT})
+  file(${WRITE} ${KCONFIG_BINARY_DIR}/Kconfig.soc.defconfig
+       "osource \"${root}/soc/$(ARCH)/*/Kconfig.defconfig\"\n"
```

Comment:
Can you provide some rationale for using 'osource' instead of checking if the file exists at configuration time and using 'source' if it does? I could see benefits to either approach and I'm curious why you chose this one.

I'm slightly concerned that files which did not exist at build system generation time can start to affect kconfig's behavior after generation, as 'osource' checks for existence when the user runs menuconfig, I think. Any issues there?

Reply:
This existing Kconfig code looks as:
```
source "$(SOC_DIR)/$(ARCH)/*/Kconfig.defconfig"
```
means that there can only be a single SOC root, which prevents support of multiple SOC roots.
Now, the `ARCH` is deduced from the board folder so that is known, but it is not enough, cause we don't know if we need to use a soc in SOC_ROOT 1, 2, 3, or `<n>`.
So we need all SOC roots in Kconfig, to be able to determine the right SOC.

Reason for using `osource`:
Example 1, someone adds a new `arm` SOC, called `mysoc`, and adds `-DSOC_ROOT=<path-to-additional-soc-root>` .
Now, as `arm` is a known arc in both `${ZEPHYR_BASE}/soc/arm` and in `<path-to-additional-soc-root>/soc/arm` then source would work.
Example 2,  User also creates `myarch`, then `${ZEPHYR_BASE}/soc/myarch` will fail if using `source` cause that folder doesn't exist, but `<path-to-additional-soc-root>/soc/myarch` works with source.
Example 2b, The opposite of 2, adding additional `arm` soc, but user decides to build for an `xtensa` soc.

But using `osource` then all cases above passes, for example `<path-to-additional-soc-root>/soc/${ARCH}`, even though `xtensa` is not present (it's present in `${ZEPHYR_BASE}/soc/xtensa`).
This means we allow users to always add additional SOC_ROOTs, even when building for an arch not present at that SOC root.
This is especially useful when allowing Zephyr modules with SOC root.

We could potentially check for existence of just any random `Kconfig.defconfig` file under `<root>/soc/${ARCH}/*` and not add the line if no file at all if found.
But that would move additional Kconfig logic into CMake, and not have it in Kconfig.
I find this knowledge to be Kconfig specific knowledge, and thus prefer it in Kconfig.
(I would have preferred an ability in Kconfig to loop a list of folders, but that is not possible, hence the reason to generate additional Kconfig files in CMake)

Also, using `osource` keeps the behavior consistent with today.
If you run:
```
cmake ... -Bbuild .
# build/zephyr/.config has been generated
mkdir ${ZEPHYR_BASE}/soc/arm/my_soc
echo 'config SOME_SETTING\n    default "myval"' > ${ZEPHYR_BASE}/soc/arm/my_soc/Kconfig.defconfig
```
then that file was not used for the first `.config` generation, but will be taken into consideration on later Kconfig invocation (including menuconfig), due the the `*` globbing:
```
source "$(SOC_DIR)/$(ARCH)/*/Kconfig.defconfig"
```

Using `osource` ensures the same behaviour, if user does:
```
cmake ... -Bbuild .
# build/zephyr/.config has been generated
# ${other-soc-root}/soc/arm does not exists, let's create it
mkdir ${other-soc-root}/soc/arm
# Now create my_soc folder
mkdir ${other-soc-root}/soc/arm/my_soc
echo 'config SOME_SETTING\n    default "myval"' > ${ZEPHYR_BASE}/soc/arm/my_soc/Kconfig.defconfig
```

If we had not added `osource \"${other-soc-root}/soc/$(ARCH)/*/Kconfig.defconfig\"\n` to the Kconfig in first place, then the behavior would be different depending on whether just a single `Kconfig.defconfig`  existed somewhere or not.

Reply:
> Also, using osource keeps the behavior consistent with today.

This more than anything else is a good reason to keep it this way IMO. Thanks for explaining.